### ISSUE TITLE
add time-slice filter

### ISSUE THREAD

**Issue by tuulos** [traildb/traildb#22]
Similar to event / column / row filter, support a filter that excludes all events outside a given time-range. Having this built-in would allow it to be used with `tdb_cons_append` for efficient materialized views.

**Comment by KenjiTakahashi** [comment]
Any plans on this one? I have some use cases where I'd like to calculate metrics on many (possibly millions) trails, but over a small time frame. Having a time filter builtin would come in handy.

**Comment by tuulos** [comment]
yes, we should add this soon

**Comment by KalenAnson** [comment]
I am not sure how you guys think this feature might best be implemented, but here's a thought:

If the filter functionality was extended to include additional operators, `<`, `>`, `<=`, and so on, the time slice functionality could leverage the existing filter framework and you could filter a time slice by using the `time` field from a trail.

I think this approach would leverage the new index functionality, but I am not entirely sure on that point.

If would tell `tdb_event_filter_add_term` the type of operator used, it currently accepts '==' and '!=' then I think this might let us just filter times (and other field values) using simple LT GT operators.

Just a thought.

**Comment by KalenAnson** [comment]
Working on a an example of this using a bit mask for the filter operator.

**Comment by tuulos** [comment]
Interesting proposal! There are pros and cons with this approach:

Pro:

 - Reuse existing APIs
 - Can support multiple time slices (kind of)
 - Can be combined with other filters

Con:

 - LT and GT operators wouldn't work with other fields, which may be confusing
 - Expressing time slices in Conjunctive Normal Form isn't very natural. For instance, consider the following sequence of timestamps with two slices:
  ```1 3 [5 6 7] 8 10 20 [25 26 30] 40 41```. You can't really use LT and GT in this case, since this doesn't work ```(time > 5 OR time > 20) AND (time < 8 OR time < 31)``` but you need to expand slices to large unions which is not very efficient.

To address the last issue, maybe there could be `tdb_event_filter_add_time_slice` which you could use to define a slice as one "term" as a part of a normal CNF query. The above example would look something like:
```
tdb_event_filter_add_time_slice(filter, 5, 8)
tdb_event_filter_add_time_slice(filter, 20, 31)
```
Probably there needs to be another parameter to control whether the slice is inclusive:exclusive etc.

Just thinking aloud. I like the idea of being able to combine this with the existing event filter functionality.

**Comment by KalenAnson** [comment]
@tuulos Agreed, there would be several pros and cons with treating time slices the same as other Trail fields and reusing the filter API for both.

To your points, Con #1 occurred to me, where a field contains a normal string value, say `derp=foo`, you are correct that LT / GT is not much help here (or any other place where we are comparing a TDB_ITEM that is the representation of a string value. Even in cases where the item is a string representation of an integer value, say `speed=200` the `200` part might not be easily compared with LT / GT operators without some type casting. (ASIDE Is there anything preventing me from storing another integer type in an event field? Say `speed`? /ASIDE)

I think that it would be very convenient and natural to use `time` intervals along side other field comparisons, even if the other field comparisons are limited to equal / not equal (PRO # 3). However I don't want the time slice functionality to suffer needlessly from performance hits because it was easy to roll this into the filter API.

Anyway, really excited about this feature :)

**Comment by tuulos** [comment]
All fields are currently simple byte-strings with no typing / semantics. You can store any data in the fields (a 64-bit int takes an 8 byte string etc) but TrailDB doesn't know that the bytes correspond to an int, so it doesn't treat them specially.

One idea that has been floating around is to eventually add support for continuous values (integers) that can be compressed efficiently. With fields like this, GT / LT operators would make more sense too.

Maybe `tdb_event_filter_add_time_slice()` is the way to go.

**Comment by tuulos** [comment]
Implemented by PR #114
